core-data/queried-data/reducer.js

```diff
@@ -0,0 +1,160 @@
+ * @param {number}         perPage     Number of items per page.
+ *
+ * @return {number[]} Merged array of item IDs.
+ */
+export function getMergedItemIds( itemIds, nextItemIds, page, perPage ) {
+	const nextItemIdsStartIndex = ( page - 1 ) * perPage;
+
+	// If later page has already been received, default to the larger known
+	// size of the existing array, else calculate as extending the existing.
+	const size = Math.max(
+		itemIds.length,
+		nextItemIdsStartIndex + nextItemIds.length
+	);
+
+	// Preallocate array since size is known.
+	const mergedItemIds = new Array( size );
+
+	for ( let i = 0; i < size; i++ ) {
+		// Preserve existing item ID except for subset of range of next items.
+		const isInNextItemsRange = (
+			i >= nextItemIdsStartIndex &&
+			i < nextItemIdsStartIndex + nextItemIds.length
+		);
+
+		mergedItemIds[ i ] = isInNextItemsRange ?
+			nextItemIds[ i - nextItemIdsStartIndex ] :
+			itemIds[ i ];
+	}
+
+	return mergedItemIds;
+}
+
+/**
+ * Higher-order reducer which returns undefined if the original reducer returns
+ * the default state value. Used in combination with `onSubKey` in avoiding
+ * assingment for undefined value to key.
+ *
+ * @param {Function} reducer Original reducer.
+ *
+ * @return {Function} Enhanced reducer.
+ */
+function withReturnUndefinedOnUnhandledDefault( reducer ) {
+	const defaultState = reducer( undefined, {} );
+	return ( state = defaultState, action ) => {
+		const nextState = reducer( state, action );
+		if ( nextState !== defaultState ) {
+			return nextState;
+		}
+	};
+}
+
+/**
+ * Reducer tracking items state, keyed by ID. Items are assumed to be normal,
+ * where identifiers are common across all queries.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Next state.
+ */
+function items( state = {}, action ) {
+	switch ( action.type ) {
+		case RECEIVE_ITEMS:
+			return {
+				...state,
+				...keyBy( action.items, 'id' ),
+			};
+	}
+
+	return state;
+}
+
+/**
+ * Reducer tracking queries state, keyed by stable query key. Each reducer
+ * query object includes `itemIds` and `requestingPageByPerPage`.
+ *
+ * @param {Object} state  Current state.
+ * @param {Object} action Dispatched action.
+ *
+ * @return {Object} Next state.
+ */
+const queries = flowRight( [
+	// Limit to matching action type so we don't attempt to replace action on
+	// an unhandled action.
+	ifMatchingAction( ( action ) => 'query' in action ),
+
+	// Inject query parts into action for use both in `onSubKey` and reducer.
+	replaceAction( ( action ) => ( {
+		...action,
+		...getQueryParts( action.query ),
+	} ) ),
+
+	// Queries shape is shared, but keyed by query `stableKey` part. Original
+	// reducer tracks only a single query object.
+	onSubKey( 'stableKey' ),
+
+	// Since query is optional for handled actions, avoid tracking empty object
+	// for default query on initialization.
+	withReturnUndefinedOnUnhandledDefault,
```

Comment:
`withReturnUndefinedOnUnhandledDefault` seems very Java to me (aka I have no idea what it does in reading it).

Reply:
Is it better or worse than [`com.seriouscompany.business.java.fizzbuzz.packagenamingpackage.interfaces.strategies. SystemOutFizzBuzzOutputStrategy`](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition/blob/master/src/main/java/com/seriouscompany/business/java/fizzbuzz/packagenamingpackage/impl/strategies/SystemOutFizzBuzzOutputStrategy.java) ? 😄

Reply:
I ended up removing `withReturnUndefinedOnUnhandledDefault`. It was needed because while `ifMatchesAction` is meant to allow the reducer to be run conditionally, we _must_ still handle the initialization case. The other option is to provide a default value as an argument of `ifMatchesAction`, but this felt redundant when the reducer should be able to handle this on its own. Instead, in 623bad6 I remove support for optional `query` from `getQueryParts` and test for the property in the action for transforming via `replaceAction`. Thus, we get the same end-result.